executor/src/witgen/jit/interpreter.rs

```diff
@@ -0,0 +1,391 @@
+                }
+            }
+        });
+
+        Self {
+            first_column_id,
+            column_count,
+            vars,
+            actions,
+        }
+    }
+
+    // Execute the machine effects for the given the parameters
+    pub fn call<Q: QueryCallback<T>>(&self, params: WitgenFunctionParams<'_, T>) {
+        let known = known_to_slice(self.column_count, params.known, params.data.len);
+        let data = params.data.into_mut_slice();
+        let pparams = params.params.into_mut_slice();
+
+        let mut vars = vec![None; self.vars];
+
+        let mut eval_stack = vec![];
+        for action in &self.actions {
+            match action {
+                InterpreterAction::AssignExpression(idx, e) => {
+                    let val = self.evaluate_expression(&mut eval_stack, &vars, e);
+                    assert!(vars[*idx].replace(val).is_none());
+                }
+                InterpreterAction::ReadCell(idx, c) => {
+                    assert!(vars[*idx]
+                        .replace(
+                            data[index(
+                                self.first_column_id,
+                                self.column_count,
+                                params.row_offset,
+                                c.row_offset,
+                                c.id,
+                            )]
+                        )
+                        .is_none())
+                }
+                InterpreterAction::ReadParam(idx, i) => {
+                    assert!(vars[*idx].replace(get_param(pparams, *i)).is_none());
+                }
+                InterpreterAction::WriteCell(idx, c) => {
+                    set(
+                        self.first_column_id,
+                        self.column_count,
+                        data,
+                        params.row_offset,
+                        c.row_offset,
+                        c.id,
+                        vars[*idx].unwrap(),
+                    );
+                }
+                InterpreterAction::WriteParam(idx, i) => {
+                    set_param(pparams, *i, vars[*idx].unwrap());
+                }
+                InterpreterAction::WriteKnown(c) => {
+                    set_known(
+                        self.first_column_id,
+                        self.column_count,
+                        known,
+                        params.row_offset,
+                        c.row_offset,
+                        c.id,
+                    );
+                }
+                InterpreterAction::MachineCall(result_vars, id, arguments) => {
+                    let mutable_state =
+                        unsafe { &*(params.mutable_state as *const MutableState<T, Q>) };
+
+                    let mut arg_values: Vec<_> = arguments
+                        .iter()
+                        .map(|a| match a {
+                            MachineCallArgument::Unknown(v) => (Some(v), Default::default()),
```

Comment:
Is this `Some(v)` used anywhere?

Do you think there is a way to directly store a `&mut` of the variable inside `args` instead of having to construct this second `arg_values` vector?

Reply:
changed things around a bit here: the knowns/unknowns are transformed into variable indexes (known arguments generate a previous assignment action)

for the second bit, i managed to do it, but the compiler didn't like it (needed a bit of unsafe)